source/grid/manifold.cc

```diff
@@ -157,61 +157,53 @@ normal_vector (const Triangulation<3, 3>::face_iterator &face,
 {
   const int spacedim=3;
 
-  // Compute distances from p to vertices
-  std::array<std::pair<double,unsigned int>, GeometryInfo<3>::vertices_per_face> distances;
-  for (unsigned int i=0; i<GeometryInfo<3>::vertices_per_face; ++i)
-    distances[i] = std::make_pair((p-face->vertex(i)).norm_square(),i);
-
-  // Sort the distances and figure out, which vertices to use
-  std::sort(distances.begin(),distances.end());
-  unsigned int first_index, second_index, fallback_index;
-
-  const double center_distance = (p-face->center()).norm_square();
-
-  // If we are closer to the center than any vertex, use closest vertices
-  if (center_distance < distances[0].first)
-    {
-      first_index = distances[0].second;
-      second_index = distances[1].second;
-      fallback_index = distances[2].second;
-    }
-  // Otherwise use vertices further away
-  else
-    {
-      first_index = distances[1].second;
-      second_index = distances[2].second;
-
-      // If we are very close to a vertex use the farthest vertex as fallback,
-      // otherwise use closest vertex.
-      if (distances[0].first < 1e4 * std::numeric_limits<double>::epsilon() * distances[1].first)
-        fallback_index = distances[3].second;
-      else
-        fallback_index = distances[0].second;
-    }
+  const std::array<Point<spacedim>, 4> vertices
+  {{face->vertex(0), face->vertex(1), face->vertex(2), face->vertex(3)}};
+  const std::array<double, 4> distances
+  {{vertices[0].distance(p), vertices[1].distance(p), vertices[2].distance(p), vertices[3].distance(p)}};
+  double max_distance = distances[0];
+  for (unsigned int i=1; i<4; ++i)
+    max_distance = std::max(max_distance, distances[i]);
+
+  // We need to find two tangential vectors to the given point p, but we do
+  // not know how the point is oriented against the face. We guess the two
+  // directions by assuming a flat topology and take the two directions that
+  // indicate the angle closest to a perpendicular one (i.e., cos(theta) close
+  // to zero).
+  double cos_angle = 1.2;
+  unsigned int first_index = 0, second_index = 1;
+  for (unsigned int i=0; i<3; ++i)
+    if (distances[i] > 1e-8*max_distance)
+      for (unsigned int j=i+1; j<4; ++j)
+        if (distances[j] > 1e-8*max_distance)
+          {
+            const double new_angle = (p-vertices[i]) * (p-vertices[j]) /
+                                     (distances[i]*distances[j]);
+            // include safety factor to avoid trouble with roundoff
+            if (std::abs(new_angle) < 0.99 * std::abs(cos_angle))
```

Comment:
I think I've convinced myself that we must always enter this `if` block at least once. If so, maybe it would be worthwhile initializing `first_index` and `second_index` above to `numbers::invalid_unsigned_int` to err on the safe side.

Reply:
Good idea.